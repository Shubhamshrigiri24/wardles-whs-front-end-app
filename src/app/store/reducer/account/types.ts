import { TActionStatus } from "app/store/types";
import { ACCOUNT_ACTIONS } from "app/store/reducer/account/constants";

export type TAccount = {
  dob: string;
  email: string;
  firstName: string;
  gender: string;
  lastName: string;
  phone: string;
  updateTime: string;
  userId: string;
  addressline1: string;
  addressline2: string;
  agreeMarketing: boolean;
  city: string;
  gpDetails: string;
  postcode: string;
};

export type State = {
  data: TAccount | null;
  status: TActionStatus;
};

export type SetAccountDataActionType = {
  type: typeof ACCOUNT_ACTIONS.SET_ACCOUNT;
  payload: TAccount;
};

export type SetAccountStatusActionType = {
  type: typeof ACCOUNT_ACTIONS.SET_ACCOUNT_STATUS;
  payload: TActionStatus;
};

export type AccountActionTypes =
  | SetAccountDataActionType
  | SetAccountStatusActionType;
