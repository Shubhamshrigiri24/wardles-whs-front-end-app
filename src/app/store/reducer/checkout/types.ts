import { RSAA } from "redux-api-middleware";
import { PostcodeLookupAddress } from "../shared";

export interface State {
  customerDetails: CustomerDetails;
  confirmedPayment: boolean;
  braintreeToken: string;
  paymentErrorMessage: string;
  postCodeLookUpAddresses: PostcodeLookupAddress[];
  deliveryDetails: DeliveryDetails;
}

export interface CustomerDetails {
  title?: string;
  firstName: string;
  lastName: string;
  line1: string;
  line2: string;
  city: string;
  postcode: string;
  email: string;
  phone: string;
  surgeryName?: string;
  surgeryAddress?: string;
  allowMarketing?: boolean;
}

export interface AddCustomerDetailsAction {
  type: string;
  payload: {
    customerDetails: CustomerDetails;
  };
}

export interface DeliveryDetails {
  title?: string;
  firstName: string;
  lastName: string;
  line1: string;
  line2: string;
  city: string;
  postcode: string;
}

export interface AddDeliveryDetailsAction {
  type: string;
  payload: {
    deliveryDetails: DeliveryDetails;
  };
}

export interface FetchBraintreeTokenAction {
  type: string;
  [RSAA]: {
    endpoint: string;
    method: string;
    types: [string, string, string];
  };
}

export interface FetchBraintreeTokenResponseAction {
  type: string;
  payload: {
    token: string;
  };
}

export interface PaymentFailureAction {
  type: string;
  payload: {
    response: {
      message: string | null;
    } | null;
  };
}

export interface FetchAddressFromPostcodeActionResponse {
  type: string;
  payload: {
    addresses: PostcodeLookupAddress[];
  };
}
