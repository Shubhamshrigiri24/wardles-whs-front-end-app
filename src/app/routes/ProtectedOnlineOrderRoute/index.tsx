export {
  default,
  ProtectedOnlineOrderRoute,
} from "./ProtectedOnlineOrderRoute";
