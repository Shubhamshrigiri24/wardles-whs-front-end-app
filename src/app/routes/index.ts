export { default, OnlineRoutes } from "./OnlineRoutes";
