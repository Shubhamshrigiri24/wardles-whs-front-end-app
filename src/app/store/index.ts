import { combineReducers, createStore, applyMiddleware, compose } from "redux";
import { apiMiddleware } from "redux-api-middleware";
import thunk from "redux-thunk";

import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";

import checkoutReducer from "./reducer/checkout";
import onlineReducer from "./reducer/online";
import orderService from "./reducer/order";
import accountService from "./reducer/account";
import subscriptions from "./reducer/subscriptions";
import orders from "./reducer/orders";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["checkout", "online", "subscriptions", "orders"],
};

const checkoutPersistConfig = {
  key: "checkout",
  storage,
  blacklist: ["braintreeToken", "confirmedPayment"],
};

const orderPersistConfig = {
  key: "order",
  storage,
};

const accountPersistConfig = {
  key: "account",
  storage,
};

const checkout = persistReducer(checkoutPersistConfig, checkoutReducer);
const order = persistReducer(orderPersistConfig, orderService);
const account = persistReducer(accountPersistConfig, accountService);

const online = persistReducer(
  {
    key: "online",
    blacklist: ["onlineServices", "productTokenValid"],
    storage,
  },
  onlineReducer
);

const rootReducers = combineReducers({
  checkout,
  online,
  order,
  account,
  subscriptions,
  orders,
});

const persistedReducer = persistReducer(persistConfig, rootReducers);

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(
  persistedReducer,
  composeEnhancers(applyMiddleware(apiMiddleware, thunk))
);

export const persistor = persistStore(store);
