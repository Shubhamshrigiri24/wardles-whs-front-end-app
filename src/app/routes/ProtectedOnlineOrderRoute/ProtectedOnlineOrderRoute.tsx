import React, { useState, useEffect, useCallback } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Redirect, useParams, useLocation, useHistory } from "react-router-dom";
import {
  OnlineServiceParamTypes,
  OnlineServiceIds,
} from "app/store/reducer/online/types";
import {
  setIsConsultationValidForOrder,
  makeSetConsultation,
} from "app/store/reducer/online/actions";

import { getOnlineIsConsultationValidForOrder } from "app/store/selectors";
import { getProduct, getCustomer } from "app/store/reducer/order/selectors";
import {
  OrderProduct,
  OrderCustomerDetails,
} from "app/store/reducer/order/types";
import { useLocationQuery } from "utils/utils";
import ConsultationAPI from "utils/api/ConsultationAPI";
import { analytics } from "@welldigital/ui-common";
import { eventsBuilder } from "utils/events";

export interface ProtectedOnlineOrderRouteProps {
  requiredStepsIds: RequiredSteps[];
  basename?: string;
}

export enum RequiredSteps {
  CONSULTATION = "consultation",
  PRODUCT_SELECTION = "product-selection",
  CUSTOMER_DETAILS = "customer-details",
}

type ValidatorsByStepType = {
  [key in RequiredSteps]: (step: Step) => boolean;
};

// A dict with validator functions for each step
const validatorsByStep: ValidatorsByStepType = {
  [RequiredSteps.CONSULTATION]: ({ data }: Step) => data as boolean,
  [RequiredSteps.PRODUCT_SELECTION]: ({ data }: Step) =>
    !!(data as OrderProduct),
  [RequiredSteps.CUSTOMER_DETAILS]: (step: Step) => {
    const data = step.data as OrderCustomerDetails | null;
    return !!data?.firstName;
  },
};

type Step = {
  id: RequiredSteps;
  data: boolean | OrderProduct | OrderCustomerDetails;
  redirectUrl: string;
};

export const ProtectedOnlineOrderRoute: React.FC<ProtectedOnlineOrderRouteProps> =
  ({ requiredStepsIds, children, basename }) => {
    const { onlineServiceId } = useParams<OnlineServiceParamTypes>();
    const isEd = onlineServiceId === OnlineServiceIds.ed;
    const isConsultationValidForOrder = useSelector(
      getOnlineIsConsultationValidForOrder
    );
    const { referral, email } = useLocationQuery();
    const { pathname } = useLocation();
    const shouldVerifyReferral =
      isEd &&
      !!referral &&
      !!email &&
      pathname === "/order/ed/product-selection";
    const [isVerifyingReferral, setIsVerifyingReferral] =
      useState<boolean>(shouldVerifyReferral);
    const product = useSelector(getProduct);
    const { customerDetails } = useSelector(getCustomer);
    const history = useHistory();
    const dispatch = useDispatch();

    const verifyReferral = useCallback(async () => {
      try {
        setIsVerifyingReferral(true);
        await ConsultationAPI.validate(referral, email);

        analytics.trackEvent({
          flow: "ed",
          event: eventsBuilder.consultation.referralLink,
          metadata: {
            isCodeValid: true,
          },
        });

        dispatch(setIsConsultationValidForOrder(true));
        makeSetConsultation(dispatch)(onlineServiceId, [
          {
            number: 0,
            initialQuestion: "Pharmacist referral",
            question: "Pharmacist referral Yes",
            answer: "Checked",
            type: "bool",
          },
        ]);
        setIsVerifyingReferral(false);
      } catch (err) {
        analytics.trackEvent({
          flow: "ed",
          event: eventsBuilder.consultation.referralLink,
          metadata: {
            isCodeValid: false,
          },
        });

        history.push(`/order/${onlineServiceId}/consultation`);
      }
    }, [
      onlineServiceId,
      history,
      referral,
      email,
      dispatch,
      setIsVerifyingReferral,
    ]);

    useEffect(() => {
      if (shouldVerifyReferral) {
        verifyReferral();
      }
    }, [shouldVerifyReferral, verifyReferral]);

    if (isVerifyingReferral) return null;

    const steps: Record<string, Step> = {
      [RequiredSteps.CONSULTATION]: {
        id: RequiredSteps.CONSULTATION,
        data: isConsultationValidForOrder,
        redirectUrl: `${
          basename ? `/${basename}` : ""
        }/${onlineServiceId}/consultation`,
      },
      [RequiredSteps.PRODUCT_SELECTION]: {
        id: RequiredSteps.PRODUCT_SELECTION,
        data: product,
        redirectUrl: `${
          basename ? `/${basename}` : ""
        }/${onlineServiceId}/product-selection`,
      },
      [RequiredSteps.CUSTOMER_DETAILS]: {
        id: RequiredSteps.CUSTOMER_DETAILS,
        data: customerDetails,
        redirectUrl: `${
          basename ? `/${basename}` : ""
        }/${onlineServiceId}/account-details`,
      },
    };

    const requiredSteps = requiredStepsIds.map(
      (stepId: RequiredSteps) => steps[stepId]
    );

    const failedStep = requiredSteps.find(
      (step: Step) => !validatorsByStep[step.id](step)
    );

    // for now the protection works only on "ed" flow
    if (!isEd) {
      return children as unknown as JSX.Element;
    }

    return failedStep ? (
      <Redirect to={failedStep.redirectUrl} />
    ) : (
      (children as unknown as JSX.Element)
    );
  };

export default ProtectedOnlineOrderRoute;
