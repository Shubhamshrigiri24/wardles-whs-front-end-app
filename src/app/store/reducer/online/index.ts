import { handleActions } from "redux-actions";
import {
  allOnlineServices,
  Ed,
  State,
  Pack,
  Product,
  defaultProduct,
  defaultPack,
  SET_PACK,
  SET_PRODUCT,
  SetPackAction,
  SetProductAction,
  FetchProductsResponseAction,
  FETCH_PRODUCTS_RESPONSE,
  FETCH_PRODUCTS_FAILURE,
  FETCH_PRODUCTS_REQUEST,
  FETCH_BRAINTREE_TOKEN_REQUEST,
  FETCH_BRAINTREE_TOKEN_RESPONSE,
  ON_ORDER_FAILURE,
  ON_ORDER_REQUEST,
  ON_ORDER_RESPONSE,
  ON_ORDER,
  CLEAR_ORDER_FAILURE_ERROR,
  OrderResponseAction,
  SetConsultationAction,
  SetIsConsultationValidForOrderAction,
  OnlineQuestion,
  SetOnlineServiceAction,
  OnlineService,
  SET_ONLINE_SERVICE,
  CHECK_PRODUCT_TOKEN_RESPONSE,
  CHECK_PRODUCT_TOKEN_FAILURE,
  CHECK_PRODUCT_TOKEN_REQUEST,
  CheckProductTokenResponseAction,
  SET_PRODUCT_TOKEN,
  SetProductTokenAction,
  INVALIDATE_PRODUCT_TOKEN,
  SEND_CONSULTATION_REQUEST,
  SEND_CONSULTATION_RESPONSE,
  SEND_CONSULTATION_FAILURE,
  SendConsultationSuccessfulResponse,
} from "./types";
import { combineReducers, Reducer } from "redux";
import { GENERIC_ERROR_MESSAGE } from "app/store/config";
import {
  FetchBraintreeTokenResponseAction,
  PaymentFailureAction,
  SET_CONSULTATION,
  SET_IS_CONSULTATION_VALID_FOR_ORDER,
  START_TRIPETTO,
} from "./types";
import { ORDER_ACTIONS } from "app/store/reducer/order/constants";

const selectedOnlineService = handleActions(
  {
    [SET_ONLINE_SERVICE]: (
      _state: OnlineService,
      action: SetOnlineServiceAction
    ) => {
      return action.payload ? action.payload.selectedOnlineService : Ed;
    },
  },
  Ed
);

const onlineServices = handleActions({}, allOnlineServices);

const selectedPack = handleActions(
  {
    [SET_PACK]: (_state: Pack, action: SetPackAction) => {
      return action.payload ? action.payload.selectedPack : defaultPack;
    },
  },
  defaultPack
);

const selectedProduct = handleActions(
  {
    [SET_PRODUCT]: (_state: Product, action: SetProductAction) => {
      return action.payload ? action.payload.selectedProduct : defaultProduct;
    },
  },
  defaultProduct
);

const products = handleActions(
  {
    [FETCH_PRODUCTS_RESPONSE]: (
      _state: Product[],
      action: FetchProductsResponseAction
    ) => {
      return action.payload ? action.payload : [];
    },
    [FETCH_PRODUCTS_FAILURE]: () => [],
    [FETCH_PRODUCTS_REQUEST]: () => [],
    [START_TRIPETTO]: () => [],
  },
  [defaultProduct]
);

const productToken = handleActions(
  {
    [SET_PRODUCT_TOKEN]: (
      _state: string | null,
      action: SetProductTokenAction
    ) => action.payload,
    [INVALIDATE_PRODUCT_TOKEN]: (_state: string | null) => null,
  },
  null
);

const productTokenValid = handleActions(
  {
    [CHECK_PRODUCT_TOKEN_RESPONSE]: (
      _state: boolean,
      action: CheckProductTokenResponseAction
    ) => {
      return !!action.payload.status;
    },
    [CHECK_PRODUCT_TOKEN_FAILURE]: () => false,
    [CHECK_PRODUCT_TOKEN_REQUEST]: () => true,
  },
  true
);

const fetching = handleActions(
  {
    [FETCH_PRODUCTS_REQUEST]: () => true,
    [FETCH_PRODUCTS_RESPONSE]: () => false,
    [FETCH_PRODUCTS_FAILURE]: () => false,
  },
  true
);

const braintreeToken = handleActions(
  {
    [ON_ORDER]: () => "",
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => "",
    [FETCH_BRAINTREE_TOKEN_RESPONSE]: (
      _state: string,
      action: FetchBraintreeTokenResponseAction
    ) => action.payload.token,
    [START_TRIPETTO]: () => "",
  },
  ""
);

const orderErrorMessage = handleActions(
  {
    [ON_ORDER_FAILURE]: (_state: string, action: PaymentFailureAction) =>
      action?.payload?.response?.message || GENERIC_ERROR_MESSAGE,
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => "",
    [ON_ORDER]: () => "",
    [CLEAR_ORDER_FAILURE_ERROR]: () => "",
  },
  ""
);

const successfulOrder = handleActions(
  {
    [ON_ORDER_FAILURE]: () => false,
    [ON_ORDER_REQUEST]: () => false,
    [SET_CONSULTATION]: () => false,
    [ON_ORDER_RESPONSE]: () => true,
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => false,
  },
  false
);

const consultationResponses = handleActions(
  {
    [SET_CONSULTATION]: (
      _state: OnlineQuestion[],
      action: SetConsultationAction
    ) => action?.payload?.consultation || [],
    [START_TRIPETTO]: () => [],
    [ORDER_ACTIONS.CLEANUP_ORDER_FLOW_STATE]: () => [],
  },
  []
);

const consultationToken = handleActions(
  {
    [SEND_CONSULTATION_REQUEST]: () => null,
    [SEND_CONSULTATION_RESPONSE]: (
      _,
      action: SendConsultationSuccessfulResponse
    ) => action.payload ?? null,
    [SEND_CONSULTATION_FAILURE]: () => null,
    [START_TRIPETTO]: () => null,
    [ORDER_ACTIONS.CLEANUP_ORDER_FLOW_STATE]: () => null,
  },
  null
);

const isConsultationValidForOrder = handleActions(
  {
    [SET_IS_CONSULTATION_VALID_FOR_ORDER]: (
      state: boolean,
      action: SetIsConsultationValidForOrderAction
    ) => action.payload,
    [START_TRIPETTO]: () => false,
    [ORDER_ACTIONS.CLEANUP_ORDER_FLOW_STATE]: () => false,
  },
  false
);

const consultation = combineReducers({
  responses: consultationResponses,
  token: consultationToken,
  isValidForOrder: isConsultationValidForOrder,
});

const orderNumber = handleActions(
  {
    [ON_ORDER_RESPONSE]: (_state: string, action: OrderResponseAction) =>
      action?.payload?.orderNumber,
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => "",
    [ON_ORDER]: () => "",
  },
  ""
);

const reducer: Reducer<State, any> = combineReducers({
  selectedOnlineService,
  onlineServices,
  selectedPack,
  selectedProduct,
  products,
  productToken,
  productTokenValid,
  fetching,
  braintreeToken,
  successfulOrder,
  orderErrorMessage,
  consultation,
  orderNumber,
});

export default reducer;
