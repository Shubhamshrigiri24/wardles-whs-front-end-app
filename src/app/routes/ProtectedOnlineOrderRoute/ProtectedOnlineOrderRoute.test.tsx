import React from "react";
import { render } from "@testing-library/react";
import { Route, Router, Switch } from "react-router";
import { createMemoryHistory } from "history";
import "@testing-library/jest-dom/extend-expect";
import * as reactRedux from "react-redux";
import { AuthenticationProvider } from "@welldigital/ui-common/Authentication";
import { store } from "app/store";
import { Store } from "redux";
import { OnlineServiceIds as mock_OnlineServiceIds } from "app/store/reducer/online/types";

import ProtectedOnlineOrderRoute, {
  ProtectedOnlineOrderRouteProps,
  RequiredSteps,
} from "./ProtectedOnlineOrderRoute";

const history = createMemoryHistory();

const state = {
  online: {
    consultation: {
      isValidForOrder: true,
    },
  },
  order: {
    customer: {
      customerDetails: {
        firstName: "John",
      },
    },
    product: {
      orderId: "123",
      id: "cialis",
      name: "Cialis",
      packs: [
        {
          itemsPerPack: 8,
          label: "8 tablets",
          name: "Cialis",
          price: 76.99,
          pricePerUnit: 9.62,
          sku: "CI60427085",
        },
      ],
      service: "ed",
      variant: "10 mg",
      subscription: false,
      image: "/static/media/cialis-10mg.123cf61a.png",
    },
  },
};

jest.mock("react-router-dom", () => {
  const { useParams, ...rest } = jest.requireActual("react-router-dom");
  return {
    ...rest,
    useParams: () => {
      return {
        onlineServiceId: mock_OnlineServiceIds.ed,
      };
    },
  };
});

let useSelectorMock = jest
  .spyOn(reactRedux, "useSelector")
  .mockImplementation((callback) => callback(state));

beforeEach(() => {
  useSelectorMock.mockClear();
});

const makeProps = (
  props: Partial<ProtectedOnlineOrderRouteProps>
): ProtectedOnlineOrderRouteProps => {
  return {
    basename: "order",
    requiredStepsIds: [
      RequiredSteps.CONSULTATION,
      RequiredSteps.PRODUCT_SELECTION,
      RequiredSteps.CUSTOMER_DETAILS,
    ],
    ...props,
  };
};

const PATH_NAMES = {
  consultation: `/order/${mock_OnlineServiceIds.ed}/consultation`,
  productSelection: `/order/${mock_OnlineServiceIds.ed}/product-selection`,
  accountDetails: `/order/${mock_OnlineServiceIds.ed}/account-details`,
  payment: `/order/${mock_OnlineServiceIds.ed}/payment`,
};

function renderWithReduxAndAuth(ui: JSX.Element, { store }: { store: Store }) {
  return render(
    <reactRedux.Provider store={store}>
      <AuthenticationProvider loginTarget={"online-services-web"}>
        {ui}
      </AuthenticationProvider>
    </reactRedux.Provider>
  );
}

describe("<ProtectedOnlineRoute />", () => {
  it("should match snapshot if no failed required step", () => {
    const props = makeProps({});
    const { container } = renderWithReduxAndAuth(
      <Router history={history}>
        <Switch>
          <Route path={PATH_NAMES.payment}>
            <ProtectedOnlineOrderRoute {...props}>
              <div data-testid={"protected/page"}>Payment page</div>
            </ProtectedOnlineOrderRoute>
          </Route>
        </Switch>
      </Router>,
      { store }
    );
    history.push(PATH_NAMES.payment);
    expect(history.location.pathname).toEqual(PATH_NAMES.payment);

    expect(container).toMatchSnapshot();
  });

  it("should redirect to consultation from product selection if consultation is not valid", () => {
    const props = makeProps({
      requiredStepsIds: [RequiredSteps.CONSULTATION],
    });

    useSelectorMock = jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((callback) =>
        callback({
          ...state,
          online: {
            consultation: {
              isValidForOrder: false,
            },
          },
        })
      );

    renderWithReduxAndAuth(
      <Router history={history}>
        <Switch>
          <Route path={PATH_NAMES.productSelection}>
            <ProtectedOnlineOrderRoute {...props}>
              <div data-testid={"protected/page"}>Product selection page</div>
            </ProtectedOnlineOrderRoute>
          </Route>
        </Switch>
      </Router>,
      { store }
    );
    history.push(PATH_NAMES.productSelection);
    expect(history.location.pathname).toEqual(PATH_NAMES.consultation);
  });

  it("should redirect to account details from payment if customer details are invalid", () => {
    const props = makeProps({
      requiredStepsIds: [
        RequiredSteps.CONSULTATION,
        RequiredSteps.PRODUCT_SELECTION,
        RequiredSteps.CUSTOMER_DETAILS,
      ],
    });

    useSelectorMock = jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((callback) =>
        callback({
          ...state,
          order: {
            ...state.order,
            customer: {
              customerDetails: null,
            },
          },
        })
      );

    renderWithReduxAndAuth(
      <Router history={history}>
        <Switch>
          <Route path={PATH_NAMES.payment}>
            <ProtectedOnlineOrderRoute {...props}>
              <div data-testid={"protected/page"}>Payment page</div>
            </ProtectedOnlineOrderRoute>
          </Route>
        </Switch>
      </Router>,
      { store }
    );

    history.push(PATH_NAMES.payment);
    expect(history.location.pathname).toEqual(PATH_NAMES.accountDetails);
  });

  it("should redirect to product selection from account details if product is invalid", () => {
    const props = makeProps({
      requiredStepsIds: [
        RequiredSteps.CONSULTATION,
        RequiredSteps.PRODUCT_SELECTION,
      ],
    });

    useSelectorMock = jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((callback) =>
        callback({
          ...state,
          order: {
            ...state.order,
            product: null,
          },
        })
      );

    renderWithReduxAndAuth(
      <Router history={history}>
        <Switch>
          <Route path={PATH_NAMES.accountDetails}>
            <ProtectedOnlineOrderRoute {...props}>
              <div data-testid={"protected/page"}>Payment page</div>
            </ProtectedOnlineOrderRoute>
          </Route>
        </Switch>
      </Router>,
      { store }
    );

    history.push(PATH_NAMES.accountDetails);
    expect(history.location.pathname).toEqual(PATH_NAMES.productSelection);
  });

  it("should redirect to first found invalid step from payment", () => {
    const props = makeProps({});
    useSelectorMock = jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((callback) =>
        callback({
          ...state,
          order: {
            ...state.order,
            product: null,
          },
        })
      );

    renderWithReduxAndAuth(
      <Router history={history}>
        <Switch>
          <Route path={PATH_NAMES.payment}>
            <ProtectedOnlineOrderRoute {...props}>
              <div data-testid={"protected/page"}>Payment page</div>
            </ProtectedOnlineOrderRoute>
          </Route>
        </Switch>
      </Router>,
      { store }
    );
    history.push(PATH_NAMES.payment);
    expect(history.location.pathname).toEqual(PATH_NAMES.productSelection);
  });
});
