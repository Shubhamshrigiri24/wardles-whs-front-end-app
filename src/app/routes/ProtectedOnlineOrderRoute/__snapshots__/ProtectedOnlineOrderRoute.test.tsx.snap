// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProtectedOnlineRoute /> should match snapshot if no failed required step 1`] = `
<div>
  <div
    data-testid="protected/page"
  >
    Payment page
  </div>
</div>
`;
